frontend/src/main/scala/bloop/engine/tasks/ScalaNative.scala

```diff
@@ -0,0 +1,88 @@
+package bloop.engine.tasks
+
+import java.net.URLClassLoader
+import java.nio.file.{Files, Path}
+
+import bloop.{DependencyResolution, Project}
+import bloop.cli.ExitStatus
+import bloop.engine.State
+import bloop.exec.Process
+import bloop.io.AbsolutePath
+import bloop.logging.Logger
+
+import monix.eval.Task
+
+object ScalaNative {
+
+  /**
+   * Compile down to native binary using Scala Native's toolchain.
+   *
+   * @param project The project to link
+   * @param entry   The fully qualified main class name
+   * @param logger  The logger to use
+   * @return The absolute path to the native binary.
+   */
+  def nativeLink(project: Project, entry: String, logger: Logger): AbsolutePath = {
+
+    initializeToolChain(logger)
+
+    val bridgeClazz = nativeClassLoader.loadClass("bloop.scalanative.NativeBridge")
+    val paramTypes = classOf[Project] :: classOf[String] :: classOf[Logger] :: Nil
+    val nativeLinkMeth = bridgeClazz.getMethod("nativeLink", paramTypes: _*)
+
+    // The Scala Native toolchain expects to receive the module class' name
+    val fullEntry = if (entry.endsWith("$")) entry else entry + "$"
+
+    AbsolutePath(nativeLinkMeth.invoke(null, project, fullEntry, logger).asInstanceOf[Path])
+  }
+
+  /**
+   * Link `project` to a native binary and run it.
+   *
+   * @param state The current state of Bloop.
+   * @param project The project to link.
+   * @param cwd     The working directory in which to start the process.
+   * @param main    The fully qualified main class name.
+   * @param args    The arguments to pass to the program.
+   * @return A task that links and run the project.
+   */
+  def run(state: State,
+          project: Project,
+          cwd: AbsolutePath,
+          main: String,
+          args: Array[String]): Task[State] = Task {
+    import scala.collection.JavaConverters.propertiesAsScalaMap
```

Comment:
Could you rebase this on top of https://github.com/scalacenter/bloop/pull/449? Also, can we not reuse as much as possible the stock implementation of `run`? This seems to boilerplate-y :smile: